Cython/Compiler/Nodes.py

```diff
@@ -6983,14 +6983,33 @@ def generate_execution_code(self, code):
             return
 
         value = self.value
-        if self.return_type.is_pyobject:
-            code.put_xdecref(Naming.retval_cname, self.return_type)
+        if value:
+            value.generate_evaluation_code(code)
+
+        if self.return_type.needs_refcounting:
+            code.putln("{")
+            code.putln(f"{self.return_type.declaration_code(Naming.quick_temp_cname)};")
+
+        if self.in_parallel:
+            # If we have the GIL we can rely on it for locking (unless in freethreading).
+            # Where we use a critical section, we do our best to keep the contents as
+            # simple as possible.
+            code.putln_openmp(
+                "#pragma omp critical(__pyx_returning)",
+                additional_tests=(
+                    "CYTHON_COMPILING_IN_CPYTHON_FREETHREADING" if not self.in_nogil_context
+                    else None))
+        code.putln("{")
+
+        if self.return_type.needs_refcounting:
+            # decref must come after assignment because it can trigger
+            # arbitrary code execution (and thus release the GIL).
+            code.putln(f"{Naming.quick_temp_cname} = {Naming.retval_cname};")
             if value and value.is_none:
                 # Use specialised default handling for "return None".
                 value = None
```

Comment:
This part was lost by moving the value evaluation up. It might not matter much, given that evaluating `None` doesn't do anything, but let's still keep the match of eval+postassign intact.